File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/search/SearchEngine.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.search;

import android.graphics.Bitmap;
import android.net.Uri;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

public class SearchEngine {

    // Parameters copied from nsSearchService.js
    private static final String MOZ_PARAM_LOCALE = "\\{moz:locale\\}";
    private static final String MOZ_PARAM_DIST_ID = "\\{moz:distributionID\\}";
    private static final String MOZ_PARAM_OFFICIAL = "\\{moz:official\\}";

    // Supported OpenSearch parameters
    // See http://opensearch.a9.com/spec/1.1/querysyntax/#core
    private static final String OS_PARAM_USER_DEFINED = "\\{searchTerms\\??\\}";
    private static final String OS_PARAM_INPUT_ENCODING = "\\{inputEncoding\\??\\}";
    private static final String OS_PARAM_LANGUAGE = "\\{language\\??\\}";
    private static final String OS_PARAM_OUTPUT_ENCODING = "\\{outputEncoding\\??\\}";
    private static final String OS_PARAM_OPTIONAL = "\\{(?:\\w+:)?\\w+\\?\\}";

    private final String identifier;
    /* package */ String name;
    /* package */ Bitmap icon;
    /* package */ List<Uri> resultsUris;
    // We don't currently support search suggestions, however that's something that we might
    // need to support in future, moreover this is already stored in the input files that we're merely
    // moving into memory - hence we probably want to keep this field for now:
    @SuppressFBWarnings(value = "URF_UNREAD_FIELD", justification = "Needed for future versions, reflects on-disk format")
    /* package */ Uri suggestUri;

    /* package */ SearchEngine(String identifier) {
        this.identifier = identifier;
        this.resultsUris = new ArrayList<>();
    }

    public String getName() {
        return name;
    }

    public String getIdentifier() {
        return identifier;
    }

    public Bitmap getIcon() {
        return icon;
    }

    public String buildSearchUrl(final String searchTerm) {
        if (resultsUris.isEmpty()) {
            return searchTerm;
        }

        // The parse should have put the best URL for this device at the beginning of the list.
        final Uri searchUri = resultsUris.get(0);

        final String template = Uri.decode(searchUri.toString());
        return paramSubstitution(template, Uri.encode(searchTerm));
    }

    public String getBaseSearchUrl() {
        if (!resultsUris.isEmpty()) {
            return resultsUris.get(0).toString();
        } else {
            return null;
        }
    }

    /**
     * Formats template string with proper parameters. Modeled after
     * ParamSubstitution in nsSearchService.js
     *
     * @param template
     * @param query
     * @return
     */
    private String paramSubstitution(String template, String query) {
        final String locale = Locale.getDefault().toString();

        template = template.replaceAll(MOZ_PARAM_LOCALE, locale);
        template = template.replaceAll(MOZ_PARAM_DIST_ID, "");
        template = template.replaceAll(MOZ_PARAM_OFFICIAL, "unofficial");

        template = template.replaceAll(OS_PARAM_USER_DEFINED, query);
        template = template.replaceAll(OS_PARAM_INPUT_ENCODING, "UTF-8");

        template = template.replaceAll(OS_PARAM_LANGUAGE, locale);
        template = template.replaceAll(OS_PARAM_OUTPUT_ENCODING, "UTF-8");

        // Replace any optional parameters
        template = template.replaceAll(OS_PARAM_OPTIONAL, "");

        return template;
    }

}

File: mobile/android/android-components/components/browser/search/src/main/java/mozilla/components/browser/search/SearchEngine.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.components.browser.search

import android.graphics.Bitmap
import android.net.Uri
import android.text.TextUtils
import java.util.Locale

/**
 * A data class representing a search engine.
 */
class SearchEngine internal constructor(
    val identifier: String,
    val name: String,
    val icon: Bitmap,
    private val resultsUris: List<Uri>,
    private val suggestUri: Uri? = null
) {
    val canProvideSearchSuggestions: Boolean = suggestUri != null

    init {
        if (resultsUris.isEmpty()) {
            throw IllegalArgumentException("Results uri list should not be empty!")
        }
    }

    /**
     * Builds a URL to search for the given search terms with this search engine.
     */
    fun buildSearchUrl(searchTerm: String): String {
        // The parse should have put the best URL for this device at the beginning of the list.
        val searchUri = resultsUris[0]

        return buildURL(searchUri, searchTerm)
    }

    /**
     * Builds a URL to get suggestions from this search engine.
     */
    fun buildSuggestionsURL(searchTerm: String): String? {
        val suggestUri = suggestUri ?: return null
        return buildURL(suggestUri, searchTerm)
    }

    private fun buildURL(uri: Uri, searchTerm: String): String {
        val template = Uri.decode(uri.toString())
        val urlWithSubstitutions = paramSubstitution(template, Uri.encode(searchTerm))
        return normalize(urlWithSubstitutions) // User-entered search engines may need normalization.
    }

    /**
     * Formats template string with proper parameters. Modeled after ParamSubstitution in nsSearchService.js
     */
    private fun paramSubstitution(template: String, query: String): String {
        var result = template
        val locale = Locale.getDefault().toString()

        result = result.replace(MOZ_PARAM_LOCALE, locale)
        result = result.replace(MOZ_PARAM_DIST_ID, "")
        result = result.replace(MOZ_PARAM_OFFICIAL, "unofficial")

        result = result.replace(OS_PARAM_USER_DEFINED, query)
        result = result.replace(OS_PARAM_INPUT_ENCODING, "UTF-8")

        result = result.replace(OS_PARAM_LANGUAGE, locale)
        result = result.replace(OS_PARAM_OUTPUT_ENCODING, "UTF-8")

        // Replace any optional parameters
        result = result.replace(OS_PARAM_OPTIONAL.toRegex(), "")

        return result
    }

    companion object {
        // We are using string concatenation here to avoid the Kotlin compiler interpreting this
        // as string templates. It is possible to escape the string accordingly. But this seems to
        // be inconsistent between Kotlin versions. So to be safe we avoid this completely by
        // constructing the strings manually.

        // Parameters copied from nsSearchService.js
        private const val MOZ_PARAM_LOCALE = "{" + "moz:locale" + "}"
        private const val MOZ_PARAM_DIST_ID = "{" + "moz:distributionID" + "}"
        private const val MOZ_PARAM_OFFICIAL = "{" + "moz:official" + "}"

        // Supported OpenSearch parameters
        // See http://opensearch.a9.com/spec/1.1/querysyntax/#core
        private const val OS_PARAM_USER_DEFINED = "{" + "searchTerms" + "}"
        private const val OS_PARAM_INPUT_ENCODING = "{" + "inputEncoding" + "}"
        private const val OS_PARAM_LANGUAGE = "{" + "language" + "}"
        private const val OS_PARAM_OUTPUT_ENCODING = "{" + "outputEncoding" + "}"
        private const val OS_PARAM_OPTIONAL = "\\{" + "(?:\\w+:)?\\w+?" + "\\}"

        private fun normalize(input: String): String {
            val trimmedInput = input.trim { it <= ' ' }
            var uri = Uri.parse(trimmedInput)

            if (TextUtils.isEmpty(uri.scheme)) {
                uri = Uri.parse("http://$trimmedInput")
            }

            return uri.toString()
        }
    }

    override fun toString(): String = "SearchEngine($identifier)"
}
